content/articles/django-custom-template-tags-and-filters/index.md

```diff
@@ -0,0 +1,258 @@
+
+{% for contact in contacts %}
+<div class="flex-box-container">
+    <div style="background-color: {{contact.name|get_color}};"><p >{{contact.name|upper}}</p></div>
+    <div><b>Created:</b> {{ contact.date }}</div>
+</div>
+{% endfor %}
+
+{% endblock content %}
+```
+
+The resulting view will look as shown below:
+
+![Contact App](/engineering-education/Django-custom-template-tags-and-filters/image2.png)
+
+The `upper ` filter capitalizes the contact name while the `get_color ` filter sets a new color to the div.
+
+Likewise, you can create your own custom filters to cater for whatever needs you may have for your templates.
+
+You should consider the names supplied to your views. They should match those used with `@register.filter()`. For instance, using `color` instead of `get_color` would result to an error. You may also have to consider the auto-escaping behaviour of Django with the filter. You may look up on Django documentation to learn more.
+
+### Writing custom template tags
+Template tags are more advanced than filters. They are powerful in that they can process any data and made available to any template regardless of view being executed. Django provides two commonly used helper functions to create tags.
+
+1. Simple tags
+2. Inclusive tags
+
+We are going to implement them in our `custom_tags.py` file.
+
+### Simple tags
+These tags take any number of arguments and return a result after some processing. The simple tag is a function that takes one or more arguments, wraps it in a render function and registers it with the template system.
+
+We are going to do two examples on the simple tags. Our first simple tag will count the number of contacts we have. The second function will get the current date in `d/m/y` format.
+
+The code in custom_tags.py is as follows:
+
+```python
+from Django import template
+import datetime
+from contact.models import Contact
+
+register = template.Library()
+
+
+@register.filter(name='upper')
+def upper(value):
+  return value.upper()
+@register.filter(name='modify_name')
+def modify_name(value):
+    return value.title()
+@register.filter(name='get_color')
+def color(value):
+    if value:
+        return "#FF0000"
+    
+@register.simple_tag(name='my_contacts')
+def my_contacts():
+    return Contact.objects.all().count()
+@register.simple_tag(name='current_date')
+def current_date(format):
+    return datetime.datetime.now().strftime(format)
+```
+
+Change the html to this:
+
+```html
+{% load custom_tags %}
+{% block content %}
+<h3>You have {% my_contacts %} contacts</h3>
+{% for contact in contacts %}
+<div class="flex-box-container">
+    <div style='background-color: {{contact.name|get_color}};'><p >{{contact.name|upper}}</p></div>
+    <div><b>Created:</b>  {% current_date "%d/%m/%Y" %} </div>
+</div>
+{% endfor %}
+
+{% endblock content %}
+```
+
+The resulting template would look as shown below:
+
+![Contact App](/engineering-education/Django-custom-template-tags-and-filters/image3.png)
+
```

Comment:
Use descriptive image names and alt text for your images. Please rename your images and change the alt text for each image.

Reply:
Done